src/passes/OptimizeInstructions.cpp

```diff
@@ -1738,16 +1743,99 @@ struct OptimizeInstructions
+      Expression* ll = nullptr;
+      Expression* rl = nullptr;
+      Expression* lr = nullptr;
+      Expression* rr = nullptr;
+      Binary* left = nullptr;
+      Binary* right = nullptr;
+
+      if (matches(curr->left, binary(&left, any(&ll), any(&lr))) &&
+          matches(curr->right, binary(&right, any(&rl), any(&rr))) &&
+          !effects(curr->right).hasSideEffects()) {
+        // canonicalize
+        // (x << C1) op (x << C2)
+        // to
+        // (x * (1 << C1)) op (x * (1 << C2))
+        bool eqLLRL = ExpressionAnalyzer::equal(ll, rl);
+        if (eqLLRL) {
+          Const *c1, *c2;
+          auto type = left->type;
+          if (matches(left, binary(Abstract::Shl, any(), constant(&c1)))) {
+            left->op = Abstract::getBinary(type, Abstract::Mul);
+            c1->value = Literal::makeFromInt32(1, type).shl(c1->value);
+          }
+          if (matches(right, binary(Abstract::Shl, any(), constant(&c2)))) {
+            right->op = Abstract::getBinary(type, Abstract::Mul);
+            c2->value = Literal::makeFromInt32(1, type).shl(c2->value);
+          }
+        }
+        // (x * y) op (x * z)
+        // (x * y) op (z * x)
+        if (left->op == right->op &&
+            left->op == Abstract::getBinary(left->type, Abstract::Mul)) {
+          bool eqLLRR = ExpressionAnalyzer::equal(ll, rr);
+          if (eqLLRR || eqLLRL) {
+            if (eqLLRR) {
+              // swap z and x
+              std::swap(rl, rr);
+            }
+            // => (y op z) * x
+            Builder builder(*getModule());
+            return builder.makeBinary(
+              Abstract::getBinary(curr->type, Abstract::Mul),
+              ll,
+              builder.makeBinary(curr->op, lr, rr));
+          }
+          // (z * y) op (x * y)
+          // (x * z) op (y * x)
+          bool eqLRRL = ExpressionAnalyzer::equal(lr, rl);
+          if (eqLRRL || ExpressionAnalyzer::equal(lr, rr)) {
+            if (eqLRRL) {
+              // swap y and z
+              std::swap(rl, rr);
+            }
+            // => (x op z) * y
+            Builder builder(*getModule());
+            return builder.makeBinary(
+              Abstract::getBinary(curr->type, Abstract::Mul),
+              lr,
+              builder.makeBinary(curr->op, ll, rl));
+          }
+        }
+      }
+    }
+    return nullptr;
+  }
+
   // given a binary expression with equal children and no side effects in
   // either, we can fold various things
   Expression* optimizeBinaryWithEqualEffectlessChildren(Binary* binary) {
-    // TODO add: perhaps worth doing 2*x if x is quite large?
+    auto type = binary->left->type;
     switch (binary->op) {
+      case AddInt32:
+      case AddInt64: {
+        // x + x  ==>  x * 2
+        binary->op = Abstract::getBinary(type, Abstract::Mul);
+        binary->right = LiteralUtils::makeFromInt32(2, type, *getModule());
+        return binary;
+      }
```

Comment:
Don't we do the opposite of this somewhere else?

Reply:
Yes, we have pr for opposite transform for floats: https://github.com/WebAssembly/binaryen/pull/3016
But for floats I propose this only due to it reduce size becouse float literals doesn't use LEB128 encodings and `f64.const 2.0` always takes 9 bytes